src/coreclr/jit/codegenriscv64.cpp

```diff
@@ -8081,7 +8090,7 @@ void CodeGen::genPopCalleeSavedRegisters(bool jmpEpilog)
                 unsigned(compiler->lvaOutgoingArgSpaceSize), totalFrameSize, compiler->compCalleeRegsPushed,
                 dspBool(compiler->compLocallocUsed));
 
-        if ((compiler->lvaOutgoingArgSpaceSize + (compiler->compCalleeRegsPushed << 3)) >= 2040)
+        if ((compiler->lvaOutgoingArgSpaceSize + (compiler->compCalleeRegsPushed << 3)) > 2040)
```

Comment:
Out of curiosity, could you explain what was wrong here?

BTW, same conditions are in genPushCalleeSavedRegisters, don't we need to change there as well?

Reply:
**Short answer:** It could be `2047` at most, because thats's the last offset index, that can be call 'directly' using uimm11. So If stack will have offsets greater than `2047`, we have to adjust stack pointer (`2040` was an index with an 8 bytes padding, but in fact there is no need for that padding).


**Longer answer:**
To better show that case let me focus on `JIT/jit64/gc/regress/vswhidbey/339415` test. 

In this test at some point in the epilog `compiler->lvaOutgoingArgSpaceSize = 2024` and only regs we saved are `ra` and `fp`.
That sum up to `2040` (each reg is 8 bytes long).
We need now space for `ra` and `fp` registers - we have to take it from outgoing arg space:
```C++
// 2024 & -16 = 2016
calleeSaveSPOffset = compiler->lvaOutgoingArgSpaceSize & -16;
```
Let's now calc remainingSPSize:
```c++
// 2024 - 2016 = 8
calleeSaveSPOffset = compiler->lvaOutgoingArgSpaceSize - calleeSaveSPOffset;
// 2304 - 8 = 2296
remainingSPSize    = remainingSPSize - calleeSaveSPOffset;
```
And here was the problem:
```matlab
2296 % 16 != 0
```
That check is being made while unwinding stack. Unwinder assumes, that everything is alligned to 16 bytes, but in our case it is not true. Of course we could 'artificially' add `STACK_ALIGN` before unwind, but that would cause errors while unwinding due to 'holes' in memory.

----

As for `genPushCalleeSavedRegisters` and `genCaptureFuncletPrologEpilogInfo` - it's very similar situation: we need to ensure, that stack offset can be use directly with `sd/ld`. As I mentioned earlier, index can be in `[0, 2047)`. So `2040` is a small padding to ensure a safe call of `offset + 8`, where we should put frame pointer.

Reply:
Yes, you definitely have right, I haven't catch that before, thank you!

Reply:
After @tomeksowi and @clamp03 comments, I've changed that fragment a little bit:
https://github.com/dotnet/runtime/blob/3fc407f1486e56bc42e189d7e130c9340ba8c774/src/coreclr/jit/codegenriscv64.cpp#L8091-L8105

So now it should be able to correct count `remainingSPSize`:
```C++
// compiler->lvaOutgoingArgSpaceSize = 2024
// 2024 & -16 = 2016
calleeSaveSPOffset = compiler->lvaOutgoingArgSpaceSize & -16;
// [...]
// 2304  - 2016 = 288
remainingSPSize    = totalFrameSize - calleeSaveSPOffset;
// 2024 - 2016 = 8
calleeSaveSPOffset = compiler->lvaOutgoingArgSpaceSize - calleeSaveSPOffset;
```
as you can see, `remainingSPSize` have correct alignment now: 
```matlab
288 % 16 == 0
```
I've tested that code a moment ago and it does work as expected.